File: mssp/examples/016-partial-and-complete/src/DMS/report.mjs
// What a person is shown.
//
// The one thing this report may never do is print a record count on its own,
// because that is the number three different situations share.
const pad = (value, width) => String(value).padEnd(width);

export function runs(rows) {
  const lines = ["  source          outcome   records   finished   error"];
  for (const row of rows) {
    const mark = { worked: "ok", empty: "--", partial: "~~", failed: "!!", absent: "  " }[row.outcome];
    lines.push(`  ${mark} ${pad(row.source, 13)} ${pad(row.outcome, 9)} ${pad(row.records.length, 9)} ` +
      `${pad(row.finished === null ? "-" : row.finished, 10)} ${row.error ?? ""}`.trimEnd());
  }
  return lines.join("\n");
}

export function combined(name, result, unfinished) {
  const lines = [];
  if (result.refused) {
    lines.push(`  ${name}: REFUSED - ${result.refused}`);
    lines.push(`  ${result.discarded} record(s) already in hand were discarded, and the work that ` +
      `produced them ran to completion anyway.`);
    return lines.join("\n");
  }
  lines.push(`  ${name}: ${result.records.length} record(s) kept`);
  if (unfinished > 0) {
    lines.push(`  ${unfinished} of them came from a source that did NOT finish - which is a fact ` +
      `about the records, not about the run.`);
  }
  return lines.join("\n");
}

export function whatACountWouldHaveSaid(rows) {
  const twos = rows.filter((row) => row.records.length === 2);
  const lines = ["  what a count alone would have said:"];
  for (const row of twos) {
    lines.push(`    ${pad(row.source, 15)} 2 records - and this is ${row.outcome}`);
  }
  lines.push(`    ${twos.length} sources, one number, ${new Set(twos.map((r) => r.outcome)).size} outcomes`);
  return lines.join("\n");
}

File: mssp/examples/017-finished-is-not-complete/src/DMS/report.py
"""What a person is shown.

The one thing this report may never print is an exact count of incomplete
records, because it does not have one. It has a floor. Section 5 of the island
test checks that the word "at least" is in the sentence, and that no exact
count is offered anywhere near it.
"""


def rows(runs):
    lines = ["  source             outcome   records   finished   complete"]
    mark = {"worked": "ok", "empty": "--", "partial": "~~", "failed": "!!", "absent": "  "}
    for row in runs:
        detail = f' ({row["reason"]})' if row["reason"] else ""
        lines.append(
            f'  {mark[row["outcome"]]} {row["source"]:<16} {row["outcome"]:<9} '
            f'{len(row["records"]):<9} {str(row["finished"]) if row["finished"] is not None else "-":<10} '
            f'{(row["completeness"] or "-")}{detail}')
    return "\n".join(lines)


def floor(at_least, total):
    return "\n".join([
        f"  at least {at_least} of {total} records come from a source that declared itself incomplete.",
        "  This is a FLOOR and not a count: a source that is truncated and says nothing is",
        "  reported in the same column as one that is complete.",
    ])


def two_axes(runs):
    finished = [row for row in runs if row["finished"]]
    lines = ["  what `finished` alone would have said:"]
    for row in finished:
        lines.append(f'    {row["source"]:<16} finished, no error, {len(row["records"])} records')
    lines.append(f"    {len(finished)} sources, all of them clean, and they are not all complete")
    return "\n".join(lines)
